src/malloc.rs

```diff
@@ -0,0 +1,128 @@
+#[no_mangle]
+pub unsafe extern "C" fn malloc(size: CSizeT) -> *mut u8 {
+	// size + 1 for to store the size of the allocated memory.
+	let layout = alloc::alloc::Layout::from_size_align(size + MAX_ALIGN, MAX_ALIGN).unwrap();
+	let ptr = unsafe { alloc::alloc::alloc(layout) };
+	if ptr.is_null() {
+		return ptr;
+	}
+	unsafe {
+		*(ptr as *mut CSizeT) = size;
+	}
+	unsafe { ptr.add(MAX_ALIGN) }
+}
+
+/// Rust implementation of C library function `calloc`
+///
+/// This function allocates memory to zero by calling Rust's global allocator.
+/// The allocated memory is aligned to the maximum alignment of any fundamental type.
+/// Due to the design of the rust global allocator, the size of the allocated memory is stored
+/// in the first 16 bytes of the allocated memory, prior to the returned pointer.
+#[no_mangle]
+pub unsafe extern "C" fn calloc(nmemb: CSizeT, size: CSizeT) -> *mut u8 {
+	let total_size = nmemb * size;
+	let ptr = malloc(total_size);
+	if ptr.is_null() {
+		return ptr;
+	}
+	unsafe {
+		core::ptr::write_bytes(ptr, 0, total_size);
+	}
+	ptr
+}
+
+/// Rust implementation of C library function `realloc`
+///
+/// This function reallocates memory by calling Rust's global allocator.
+/// The allocated memory is aligned to the maximum alignment of any fundamental type.
+/// Due to the design of the rust global allocator, the size of the allocated memory is stored
+/// in the first 16 bytes of the allocated memory, prior to the returned pointer.
+#[no_mangle]
+pub unsafe extern "C" fn realloc(ptr: *mut u8, size: CSizeT) -> *mut u8 {
+	if ptr.is_null() {
+		return malloc(size);
+	}
+	let old_size = unsafe { *(ptr.sub(MAX_ALIGN) as *mut CSizeT) };
+	let layout = alloc::alloc::Layout::from_size_align(old_size + MAX_ALIGN, MAX_ALIGN).unwrap();
+	let new_ptr = unsafe { alloc::alloc::realloc(ptr.sub(MAX_ALIGN), layout, size + MAX_ALIGN) };
+	if new_ptr.is_null() {
+		return new_ptr;
+	}
+	unsafe {
+		*(new_ptr as *mut CSizeT) = size;
+	}
+	unsafe { new_ptr.add(MAX_ALIGN) }
+}
+
+/// Rust implementation of C library function `free`
+///
+/// This function frees memory by calling Rust's global allocator.
+/// The allocated memory is aligned to the maximum alignment of any fundamental type.
+/// Due to the design of the rust global allocator, the size of the allocated memory is stored
+/// in the first 16 bytes of the allocated memory, prior to the returned pointer.
+#[no_mangle]
+pub unsafe extern "C" fn free(ptr: *mut u8) {
+	if ptr.is_null() {
+		return;
+	}
+	let old_size = unsafe { *(ptr.sub(MAX_ALIGN) as *mut CSizeT) };
+	let layout = alloc::alloc::Layout::from_size_align(old_size + MAX_ALIGN, MAX_ALIGN).unwrap();
+	unsafe { alloc::alloc::dealloc(ptr.sub(MAX_ALIGN), layout) };
+}
+
+#[cfg(test)]
+mod test {
+	use super::*;
+
+	#[test]
+	fn test_malloc() {
+		let ptr = unsafe { malloc(10) };
+		assert!(!ptr.is_null());
+		unsafe {
+			assert_eq!(*(ptr.sub(MAX_ALIGN) as *mut CSizeT), 10);
+		}
+		unsafe { free(ptr) };
```

Comment:
I suggest we also write to every byte in the buffer (e.g. with a rolling value like `index as u8`), so that we verify we did get as much memory as we thought we did.